### ISSUE TITLE
User can define optional features, which map to Maven's <optional> dependencies

### ISSUE THREAD

**Issue by madorb** [gradle/gradle#867]
Gradle release 3.2 has come and gone the 5 year old JIRA Issue regarding the lack of `optional` dependencies has not yet been addressed. This was somewhat addressed by `compileOnly` - though that often seems to be described as a solution for `provided` but it doesn't appear that it precisely exhibits the behavior of either of them. Presently when i mark a dependency as `compileOnly` it is not emitted in my published pom file, which is not the desired behaviour, it should be emitted as an optional dependency to ensure that any consumers can align on the necessary version.

### Expected Behavior

The `optional` configuration for dependencies should follow the same behaviour as maven, such that the dependencies are available at compile time but are reflected in published `pom.xml` files as `optional` 

### Current Behavior

The functionality does not exist outside of various 3rd party plugins such as the [spring gradle propdeps plugin](https://github.com/spring-projects/gradle-plugins/tree/master/propdeps-plugin), which of course results in these configurations not being respected in things like the maven-publish plugin. 

### Context

This is a frequent occurrence especially when building libraries for third party consumption, presently this requires every time to come up with their own particular gradle hack to allow publishing of maven artifacts.  

### Steps to Reproduce (for bugs)

N/A

### Your Environment

All gradle projects

### Additional Context
[GRADLE-1749](https://issues.gradle.org/browse/GRADLE-1749) open since 2011-08-15

**Comment by digulla** [comment]
The ["Migate from Maven to Gradle" guide mentions two plugins](https://guides.gradle.org/migrating-from-maven/#provided_scope_and_optional_dependencies) which offer this functionality: 

```
plugins {
    id 'nebula.provided-base' version '2.2.2'
    id 'nebula.optional-base' version '2.2.2'
}

```

**Comment by rohit-zabbed** [comment]
Hi Team, please look into this, I can track back and this issue for sure exists for more than 4 years. Really appreciate if it can be addressed now.

**Comment by ljacomet** [comment]
While supporting _optional_ dependencies, that are needed when given features of a component are enabled, remains something that Gradle should offer one day, the comparison to Maven `optional` dependencies does not hold.

See the conversation in #5434 around parsing and accounting for `optional` dependencies from Maven POM into Gradle resolution. In short, Maven simply ignores `optional` dependencies, never using them for conflict resolution at all. `optional` and `provided` dependencies in Maven are more documentation for users of the component.

**Comment by ljacomet** [comment]
This will be released as "feature variants" in Gradle 5.3, watch out for the release notes and documentation!

**Comment by danmana** [comment]
Here is the documentation for those having a hard time finding it:
https://docs.gradle.org/5.3.1/userguide/feature_variants.html

And the release notes:
https://docs.gradle.org/5.3.1/release-notes.html#feature-variants-(similar-to-optional-dependencies-in-maven)
